_posts/2015-12-13-testing-multithreaded-code-in-java.md

```diff
@@ -0,0 +1,154 @@
+    	throws	InterruptedException {
+        MultithreadedStressTester stressTester =
+        new MultithreadedStressTester(25000);
+
+        stressTester.stress(new Runnable() {
+            public void run() {
+                counter.inc();
+            }
+        });
+
+        stressTester.shutdown();
+
+        assertThat("final count", counter.count(),
+        equalTo(BigInteger.valueOf(stressTester.totalActionCount())));
+    }
+```
+
+The stress tester will exercise the method n loops with m threads. As our method is incrementing by one, we should see that n*m is equal to the counter.count().
+
+The interesting class is the MultithreadedStressTester though:
+
+```
+	public void stress(final Runnable action) throws InterruptedException {
+        spawnThreads(action).await();
+    }
+
+    private CountDownLatch spawnThreads(final Runnable action) {
+        final CountDownLatch finished = new CountDownLatch(threadCount);
+
+        for (int i = 0; i < threadCount; i++) {
+            executor.execute(new Runnable() {
+                public void run() {
+                    try {
+                        repeat(action);
+                    }
+                    finally {
+                        finished.countDown();
+                    }
+                }
+            });
+        }
+        return finished;
+    }
+
+    private void repeat(Runnable action) {
+        for (int i = 0; i < iterationCount; i++) {
+            action.run();
+        }
+    }
+```
+
+If you execute that test you will receive different results and sometimes it's even passing! That's because this test is not deterministic, we can't assure how the threads will interleave in every execution. If we want to be as sure as possible that this test finds the possible bug, we should increase the number of threads and iterations, but with the obvious time trade-off.
+
+You can use a more deterministic approach using [Weaver](https://github.com/google/thread-weaver). To understand how it works let's illustrate it with an example. Let's say that we have an in-memory and not thread-safe store:
+
+```
+	private final Map<Level, Scores> scoresByLevel;
+```
+
+We have some service that access to a repo that wraps that store:
+
+```
+  1	Optional<Scores> scoresFromStore = scoreRepo.findBy(score.level());
+	2       if(scoresFromStore.isPresent()){
+	3          scoreRepo.update(score.level(), score);
+	4       }else{
+	5          scoreRepo.save(score.level(), new Scores().add(score));
+	6       }
+```
+
+That service is a singleton living in a server that spawns a thread per request, so we'd like to execute that piece atomically. We could use the stress test non-deterministic approach or we could use Weaver. If we think deeply in this problem, what we want to test is every combination of the following (as an example, Thread 1 executes line 1 in moment x and Thread 2 executes line 1 in moment x, would be -> T1/1 : T2/1)
```

Comment:
If we think deeply about this problem, we realise we want to test every combination of the following (..):